Experimental Rudder implementation

nebril opened the pull request:
This change introduces the concept of Rudders - pluggable modules that
Tiller communicates with via grpc, which allow to decouple orchestration
logic from Tiller into separate service.

This commit consists of simple Rudder implementation which does exactly
the same thing as built in Tiller orchestrator - it creates all k8s
objects from provided manifest without orchestrating them.

--experimental-release flag is introduced to enable this behaviour.

This change allows to use the service and tiller outside of the cluster.
Following commits will add Rudder to helm deployment.

It is the first PR in effort to implement #1924

nebril committed:
Experimental Rudder implementation

This change introduces the concept of Rudders - pluggable modules that
Tiller communicates with via grpc, which allow to decouple orchestration
logic from Tiller into separate service.

This commit consists of simple Rudder implementation which does exactly
the same thing as built in Tiller orchestrator - it creates all k8s
objects from provided manifest without orchestrating them.

--experimental-release flag is introduced to enable this behaviour.

This change allows to use the service and tiller outside of the cluster.
Following commits will add Rudder to helm deployment.

nebril commented:
@technosophos @adamreese please share your feedback. If the implementation is correct, should this code be merged, or should the part in which Rudder is actually incorporated into helm pod also be part of this PR?

Checking out CI issues now.

technosophos commented on pkg/tiller/release_server.go:
```diff
@@ -866,8 +875,7 @@ func (s *ReleaseServer) performRelease(r *release.Release, req *services.Install
 		// regular manifests
-		b := bytes.NewBufferString(r.Manifest)
-		if err := s.env.KubeClient.Create(r.Namespace, b, req.Timeout, req.Wait); err != nil {
+		if err := s.ReleaseModule.Create(r, req, s.env); err != nil {
```
This makes me very happy. It is a very clean integration.

technosophos commented on cmd/tiller/tiller.go:
```diff
@@ -82,6 +83,7 @@ func main() {
 	p.StringVarP(&grpcAddr, "listen", "l", ":44134", "address:port to listen on")
 	p.StringVar(&store, "storage", storageConfigMap, "storage driver to use. One of 'configmap' or 'memory'")
 	p.BoolVar(&enableTracing, "trace", false, "enable rpc tracing")
+	p.BoolVar(&remoteReleaseModules, "experimental-release", false, "enable experimental release modules")
```
Glad you implemented it this way.

technosophos commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,84 @@
+
+	"k8s.io/helm/pkg/kube"
+	"k8s.io/helm/pkg/proto/hapi/release"
+	"k8s.io/helm/pkg/rudder"
```
This appears to have not been checked in.

nebril replied:
Indeed. Sorry about that! I thought there was an issue with CI building against older version of Helm :D.

technosophos commented:
@nebril Looking good! I'm pleased to see that the interface is looking really clean. I was worried that there might be a major release server refactor lurking. But it looks good.

Which version of `protoc` are you using? I'm trying to track down the origin of some of the regeneration changes that I am seeing on lots of PRs.

nebril commented:
@technosophos I am using libprotoc 3.2.0 . I don't have a lot of experience with protobuffs, so I am not sure what caused all of these binary changes there.

technosophos commented:
Okay, I just read through all of the code, and it is exactly on track. It feels like a very clean separation of concerns.

I think you can just keep on going with the upgrade/delete/rollback logic.

I'll add @adamreese and I as reviewers on the issue. I'm also going to cc @fibonacci1729, who's the resident expert on gRPC.

nebril committed:
Add experimental dockerfiles

Added dockerfiles for rudder service and tiller run with
--experimental-release flag

dshulyak commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,84 @@
+	if err != nil {
+		grpclog.Printf("error when creating release: %s", err)
+	}
+	return &release.InstallReleaseResponse{}, err
```
why response is empty?

nebril replied:
An oversight on my part.

dshulyak commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,84 @@
+func (r *ReleaseModuleServiceServer) InstallRelease(ctx context.Context, in *release.InstallReleaseRequest) (*release.InstallReleaseResponse, error) {
+	grpclog.Print("install")
+	b := bytes.NewBufferString(in.Release.Manifest)
+	err := kubeClient.Create(in.Release.Namespace, b, 500, false)
```
are you sure that we shouldn't wait here? because 500 seconds of timeout without waiting is meaningless

nebril replied:
This timeout should be derived from the request, I did not include passing parameteres from InstallReleaseRequest sent to tiller to rudder yet.

dshulyak commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,84 @@
+
+	grpclog.Print("Server starting")
+	grpcServer.Serve(lis)
+	grpclog.Print("Server started")
```
i think this is unreachable

nebril replied:
Indeed.

nebril committed:
Use helm version as rudder version

dshulyak committed:
Implement upgrade for rudder

dshulyak committed:
Update release proto

dshulyak committed:
Run make protoc to update hapi

dshulyak committed:
Use rollback method on ReleaseModule interface

thomastaylor312 commented:
Per the new developer workflow guidelines, I have removed the `wip` label and have prefaced the issue with  "WIP." Please remove the "WIP" preface when it is ready for review.

dshulyak committed:
Define both rudder port and rudder address for different needs

nebril committed:
Move rudder proto to different package, add release status

nebril committed:
ReleaseServer.ReleaseModule has Status method

Both local and remote Release module implement release status check.

jellonek commented on _proto/hapi/rudder/rudder.proto:
```diff
@@ -0,0 +1,118 @@
+
+syntax = "proto3";
+
+package hapi.services.rudder;
```
`services` there looks like reminder from some refactors, right?

jellonek commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,106 @@
+limitations under the License.
+*/
+
+package main // import "k8s.io/helm/cmd/rudder"
```
With other packages such comment have a sense, but with `main`?

jellonek commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,106 @@
+	return &rudderAPI.InstallReleaseResponse{}, err
+}
+
+// DeleteRelease is not implemented
```
Shouldn't that be implemented in this PR?

jellonek commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,106 @@
+	return nil, nil
+}
+
+// RollbackRelease is not implemented
```
Looks like implemented ;)

jellonek commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,106 @@
+// ReleaseModuleServiceServer provides implementation for rudderAPI.ReleaseModuleServiceServer
+type ReleaseModuleServiceServer struct{}
+
+// Version is not yet implemented
```
Looks like implemented to me :)

technosophos replied:
Still need to fix this comment.

nebril replied:
Done.

jellonek commented on pkg/tiller/release_modules.go:
```diff
@@ -0,0 +1,102 @@
+	return env.KubeClient.Update(target.Namespace, c, t, req.Recreate, req.Timeout, req.Wait)
+}
+
+func (m *LocalReleaseModule) Status(r *release.Release, req *services.GetReleaseStatusRequest, env *environment.Environment) (string, error) {
```
Missing docstring for exported func.

nebril replied:
done

jellonek commented on pkg/tiller/release_modules.go:
```diff
@@ -0,0 +1,102 @@
+	return env.KubeClient.Create(r.Namespace, b, req.Timeout, req.Wait)
+}
+
+func (m *LocalReleaseModule) Update(current, target *release.Release, req *services.UpdateReleaseRequest, env *environment.Environment) error {
```
Missing docstring for exported func.

nebril replied:
done

jellonek commented on pkg/tiller/release_modules.go:
```diff
@@ -0,0 +1,102 @@
+	return env.KubeClient.Update(target.Namespace, c, t, req.Recreate, req.Timeout, req.Wait)
+}
+
+func (m *LocalReleaseModule) Rollback(current, target *release.Release, req *services.RollbackReleaseRequest, env *environment.Environment) error {
```
Missing docstring for exported func.

nebril replied:
done

jellonek commented on pkg/tiller/release_modules.go:
```diff
@@ -0,0 +1,102 @@
+	return err
+}
+
+func (m *RemoteReleaseModule) Status(r *release.Release, req *services.GetReleaseStatusRequest, env *environment.Environment) (string, error) {
```
Missing docstring for exported func.

nebril replied:
done

technosophos commented:
This one is still listed as WIP -- I believe the deletion command is still waiting completion. Is there anything else?

nebril commented:
@technosophos no, just the deletion command. It's coming.

nebril committed:
Basic Rudder Delete implementation

Extracted delete specific code from ReleaseServer to external function
which is called from both Local and Remote ReleaseModules.

Made getVersionSet function from tiller package exported.

technosophos commented:
Moving this to 2.4. Sorry for my delays which caused us to have to slip this a release. But this should give us a good long "pre-release" testing cycle in 2.4.0.

technosophos commented:
Okay, the big Kube 1.6 change was done. This probably needs a quick rebase, then we can start testing it out.

nebril commented:
@technosophos rebased

nebril commented:
Rebased again, empty yaml handling in delete moved to extracted Delete function.

technosophos commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,140 @@
+
+	if len(allErrors) > 0 {
+		err = fmt.Errorf(allErrors)
+	} else {
```
I don't think you need this `else` block.

nebril replied:
Indeed. Fixed

technosophos commented on pkg/tiller/release_modules.go:
```diff
@@ -0,0 +1,168 @@
+// DeleteRelease is a helper that allows Rudder to delete a release without exposing most of Tiller inner functions
+func DeleteRelease(rel *release.Release, vs chartutil.VersionSet, kubeClient environment.KubeClient) (kept string, errs []error) {
+	manifests := relutil.SplitManifests(rel.Manifest)
+	_, files, err := sortManifests(manifests, vs, UninstallOrder)
```
Good idea moving that logic in here. That will be a very good thing in the future, since some rudders may have a different deletion order than the default implementation.

technosophos reviewed (changes requested):
I have only two very minor changes, then I'm good with merging this.

nebril committed:
Add comments to exported Rudder methods

Added make target
Lowered docker image sizes
Updated ruder.pb.go

nebril reviewed (commented):
Thanks for review @technosophos and @jellonek !

Fixed your comments, PTAL.

technosophos reviewed (approved):
Reviewed the most recent batch of changes, and I think we are good to merge. I will double-check with @adamreese today.

adamreese commented on cmd/rudder/rudder.go:
```diff
@@ -0,0 +1,138 @@
+type ReleaseModuleServiceServer struct{}
+
+// Version returns Rudder version based on helm version
+func (r *ReleaseModuleServiceServer) Version(ctx context.Context, in *rudderAPI.VersionReleaseRequest) (*rudderAPI.VersionReleaseResponse, error) {
```
Does this get called anywhere?

technosophos replied:
It is required by the proto, but we haven't added a call for it anywhere. The intention was to make this available for (a) logging, and (b) the future toggle switch that allows choosing which rudder to run for a given release.

nebril replied:
Matts comment is correct, I have nothing further to add.

adamreese commented on Makefile:
```diff
@@ -60,12 +61,20 @@ docker-binary: BINDIR = ./rootfs
 docker-binary: GOFLAGS += -a -installsuffix cgo
 docker-binary:
 	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 $(GO) build -o $(BINDIR)/tiller $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' k8s.io/helm/cmd/tiller
+	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 $(GO) build -o $(BINDIR)/rudder $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' k8s.io/helm/cmd/rudder
```
Can you move this to a different target so CI won't need to build it.

adamreese commented on rootfs/Dockerfile.experimental:
```diff
@@ -0,0 +1,24 @@
+
+ENV HOME /tmp
+
+COPY . /
```
This should only copy the tiller binary to cut the image size in half.

adamreese commented on rootfs/Dockerfile.rudder:
```diff
@@ -0,0 +1,23 @@
+
+ENV HOME /tmp
+
+COPY . /
```
This should only copy the rudder binary to cut the image size in half. _feels deja vu_

technosophos commented:
Bumping to 2.5 to clear off for today's release.

nebril commented:
@adamreese fixed the stuff you mentioned, please review :)

technosophos commented:
Re-reviewed and it looks good still.
